### PR TITLE
Contract Rebuild

### PR DESCRIPTION
**Pull request by pap1723**
Here is a Rebuild of the Contract System. More details forthcoming...

### CODE REVIEW

**Comment by pap1723** [comment]
**THIS WILL NOT BREAK YOUR SAVE GAMES**
The first part of the Contract Overhaul is complete and available for testing. There are currently 138 contracts in game. Every contract that was in before is still there, just with some small modifications which might include Prerequisites and wording of details.

Please, please, please test and report back.

## CURRENTLY INCLUDED
**MILESTONES, X-PLANES and SOUNDING ROCKETS**
There are the same as before with only two additions, a Suborbital Return contract and a First Rendezvous Contract.

**EARLY SATELLITES**
All brand new contracts that are designed to introduce the players to the different types of orbits. These include Polar, Molniya, Sun-Synchronous, Tundra and Geostationary. There are also contracts available to create an Early Communications Network of 3 or 4 satellites. These contracts become available after the First Satellite contract is completed.

**HUMAN MILESTONES**
These are very, very similar to the contracts from before.

**MOON EXPLORATION**
This has moved all of the Lunar contracts into their own contract group. This includes all of the normal stuff from before (Flyby, Impact, Orbit, Landing). This also includes an Unmanned Lunar Landing and Return mission (like Luna 16) and a Moon Rover mission (like Lunokhod 1). The Human Moon Spaceflights are included as well and includes a new mission to bring a rover to the Moon as well.

**FLYBYS**
Uncrewed flybys for all planetary bodies in RSS. No change from before, but I have standardized the contract payouts. 20% of the money is up front and the rest comes after successful completion. I have also removed the need to have a New Vessel launched, therefore if you are on a mission to Flyby Jupiter and can get an encounter with another Outer Planet, you can accept the contract and get the rewards for it.

**ORBITS**
Unlock after the first flyby of the respective planetary body.
Uncrewed orbits for all other 7 planets and Pluto.

**ATMOSPHERIC PROBES**
Unlock after the first flyby of the respective planetary body.
New missions to put a probe into the upper atmosphere of all planetary bodies with an atmosphere. Requires science to be transmitted back from the upper atmosphere.

**LANDINGS**
Unlock after the first flyby of the respective planetary body.
No real changes from before, all planetary bodies with a solid surface have a mission to land on them.

**ROVERS**
Unlock after the first flyby of the respective planetary body.
Every planetary body that can be landed on now has a contract to send a rover and to explore 3 waypoints. This requires a targeted landing, so they are definitely advanced missions.

**SAMPLE RETURN**
Unlock after landing on the respective planetary body.
These are uncrewed missions that require you to land on a planetary body, collect science and fly it back to Earth.

**SPACE STATION**
Only the first space station mission is currently available. It is unlocked after a successful completion of the First Docking contract and First EVA.

## COMING SOON™
* Space Station Contracts
  * Lunar Space Station
  * Martian Space Station
  * Crew Rotation
  * Add Power Module
  * Increase Crew Space on Station
  * Resupply Life Support
  * Deorbit Existing Station
  * Put up a New Station
* Human Exploration
  * Venus Flyby
  * Mars Flyby
  * Venus Orbit
  * Mars Orbit
  * Phobos Landing
  * Mars Landing
  * Asteroid Landing
  * Jupiter Flyby
  * Europa Landing
  * Saturn Flyby
* Surface Bases
  * Lunar Base
  * Martian Base
  * Base Crew Rotation
  * Base Expansion
  * Base Resupply

File: GameData/RP-0/Contracts/Moon Exploration/Crewed Moon Landing and Rover.cfg [hunk 1]
```diff
@@ -0,0 +1,187 @@
+CONTRACT_TYPE
+{
+	name = MoonExtendedStayCrew
+	title = Crewed Moon Landing & Rover Exploration
+	group = MoonExploration
+
+	description = Design and launch a spacecraft with at least one crew member to land on the Moon. This will be a targeted landing near a randomly generated waypoint. We will also require you to take a manned rover to explore an additional two waypoints. Once you have explored the waypoints, return safely to Earth. This mission can be completed 2 times.
```
Comment (rsparkyc) [inline GameData/RP-0/Contracts/Moon Exploration/Crewed Moon Landing and Rover.cfg]: I saw you were changing manned to crewed, but you still have one reference here

**Comment by leudaimon** [comment]
I don't know if I understood the orbital parameters for Molniya and Tundra orbits correctly, but it looks like they are restricted to be oriented with a northward apoapsis. Even though this is historically accurate, I wonder if these contracts shouldn't be more general, with a possibility of coverage of southern latitudes. Given the reason for only northern orbits is more geopolitical than for some other reason, I think it would be plausible for a southern country, such as Chile or South Africa (or even communications to Antarctica) to use these orbits with a southern orientation in an alternate history.

**Comment by pap1723** [comment]
@leudaimon you bring up a good point about the orbits. The only reason they are the way they are is because of the countries involved in using them.

I can change the Argument of Periapsis to be anything. The issue is going to be that I cannot limit it to either the North or the South. That means that the player would be able to have a contract with a Perigee near the equator, defeating the purpose of the orbit type.

Anyone else have thoughts?

**Comment by leudaimon** [comment]
Given the stability gained by using an inclination of exactly 63.4°, and the fact these orbits are mostly useful to reach contact with high latitudes in contrast with GEO that reaches low/medium latitudes, I think it makes sense to keep the arguments of periapsis to coincide with either the highest or lowest latitudes reached by an orbit with a 63.4° inclination.
Is it possible to keep the longitude of AN free, and the argument of periapsis to have two possible values referring to south or north extremes?

**Comment by pap1723** [comment]
I have updated this with the first part of the Weather Satellite Contracts. What I was going for is a progressive difficulty and necessary parameters. Here is what is available:

## WEATHER SATELLITE CONTRACTS
* **First Weather Satellite** - Just get a satellite up in orbit above 300 km Perigee and relatively circular. Similar to TIROS I from history (Available after the Atmosphere Analysis contract is completed)
* **Early Weather Satellites** - Repeatable (up to 6 times) contracts that are similar to TIROS 2 through TIROS 7. Randomly generated Parameters with a minimum Perigee from 596 km to 737 km, an Inclination between 48 and 58 degrees and a maximum Eccentricity of 0.01. (Available after the First Weather Satellite contract is completed, cannot have anymore after the Second Generation Weather Satellites are accepted)
* **Second Generation Weather Satellites** - The first of the Sun Synchronous Weather Satellites similar to the NIMBUS program. Repeatable up to 6 times. These are higher orbits with a minimum Perigee from 1000 km, an Inclination of 95 to 100 degrees and a maximum Eccentricity of 0.005 (Available after at least 3 Early Weather Satellite contracts are completed, and are no longer available when the Geostationary or Sun Synchronous contracts are accepted)
* **Geostationary Weather Satellites** - These repeatable contracts can be done up to 6 times and require the player to put a contract in a Geostationary Orbit within 11.25 degrees of a randomly generated waypoint on the Equator. (Available once you have completed 3 Second Generation Weather Satellite contracts)
* **Sun Synchronous Weather Satellites** - These repeatable contracts can be done up to 6 times and require the player to put a satellite in a defined orbit.  (Available once you have completed 3 Second Generation Weather Satellite contracts)

**Future Weather Contracts**
* Replace an Aging Geostationary Satellite

**Comment by pap1723** [comment]
Added some new contracts today. We are getting close to being done!

# New Contracts
## SPACE STATIONS
* Crew Rotation
  * Repeatable. Will not appear again for 90 days after the successful completion of the same mission
  * Requires the player to bring a new crew to the station and then to remove the old crew and bring them back to Earth
* Life Support Resupply
  * Repeatable. Searches for stations with 25% or less of their remaining Food, Water or Oxygen
  * Requires the player to bring enough Food, Water and Oxygen to completely refill the stocks of the station
* Add Crew Module
  * Repeatable. Will not appear again for 2 years after the successful completion of the same mission
  * Requires the player to bring a new module to their space station that has habitation for at least 4 crew.
* New Station Crew
  * Repeatable. Will not appear again for 6 months after the successful completion of the same mission. It searches all stations for crew capacity with 0 crew on board.
  * Requires the player to bring a new crew of at least 2 to the station.
* New Power Module
  * Repeatable. Will not appear again for 2 years after the successful completion of the same mission
  * Requires the player to bring a new module to the space station with the ability to generate at least 4.0 electric per second.

## HUMAN EXPLORATION
* Venus Flyby
  * Available after the 2nd Moon Landing and after an uncrewed Venus Flyby
  * Send at least 2 crew on a flyby of Venus and return home safely
* Mars Flyby
  * Available after the 2nd Moon Landing and after an uncrewed Mars Flyby
  * Send at least 2 crew on a flyby of Mars and return home safely
* Venus Orbit
  * Available after the 2nd Moon Landing and after an uncrewed Venus Orbit
  * Send at least 2 crew to orbit Venus and return home safely
* Mars Orbit
  * Available after the 2nd Moon Landing and after an uncrewed Mars Orbit
  * Send at least 2 crew to orbit Mars and return home safely
* Mars Landing
  * Available after the 2nd Moon Landing and after an uncrewed Mars Landing
  * Send at least 2 crew to land on Mars for 30 days and return home safely
* Phobos Landing
  * Available after the 2nd Moon Landing and after an uncrewed Phobos Landing
  * Send at least 2 crew to land on Phobos for 12 hours and return home safely
* Asteroid Landing
  * Available after the Mars Crewed Landing and after an uncrewed Asteroid Landing
  * Send at least 2 crew to land on an Asteroid and return home safely
* Mercury Flyby
  * Available after a crewed flyby of Mars and Venus and an uncrewed Mercury Flyby
  * Send at least 2 crew on a flyby of Mercury and return home safely
* Jupiter Flyby
  * Available after a crewed flyby of Mars and Venus and an uncrewed Jupiter Flyby
  * Send at least 3 crew on a flyby of Jupiter and return home safely
* Saturn Flyby
  * Available after a crewed flyby of Mars and Venus and an uncrewed Saturn Flyby
  * Send at least 3 crew on a flyby of Saturn and return home safely

**Comment by NathanKell** [comment]
Just to be clear in 'public' as it were since I've already said this on IRC: THIS IS AWESOME!

**Comment by PhineasFreak** [comment]
@pap1723 I absolutely love the new tech tree patching for the engine configs but there is room for improvement. You can easily group the engine configs of a single engine by doing:

```
@PART[*]:HAS[#engineType[Aerobee]]:FOR[zzRP-0]
{
    @MODULE[ModuleEngineConfigs]
    {
        @CONFIG[WAC-Corporal]
        {
            %techRequired = start
        }

        @CONFIG[XASR-1]
        {
            %techRequired = basicRocketry
        }

        @CONFIG[AJ10-27]
        {
            %techRequired = basicRocketry
        }
    }
}
```

So to minimize both your effort on future additions/maintenance (all respective configs in one place) and the number of MM "patching loops" (currently the Aerobee engines require 3 passes each for the 3 engine configs instead of 1 - more time is required to apply them). We can also include the costing fields to these as RO has nothing to do with either tech trees or part costs.

**Comment by pap1723** [comment]
Hey @PhineasFreak that makes good sense! Right now, the file is built using an Excel sheet that creates the Engine_Config file by basically puling each row out and making a "new entry" for it. Every engine config is on a separate row so it creates each one like that and cannot group them together.

I personally do not have enough Excel-fu to be able to group them together if it contains the same Engine Config as the master config.

**Comment by pap1723** [comment]
I wasn't done, but accidentally hit send.

This process can definitely be improved in the future and I am willing to hear all ideas on how to make it better!

**Comment by jwvanderbeck** [comment]
+1 for using procedural things like engineType.  It is almost more important to look forward to newer parts, than it is to manage existing parts.  This is part (ug) of the reason why it is so hard to move forward on expanding RO because adding a new part involves so many connections.

**Comment by PhineasFreak** [comment]
@pap1723 that is OK, it is always possible to edit them, even by hand if necessary. Besides, there is going to be a lot of work to port everything over from RO in just one step.

**Comment by leudaimon** [comment]
I don't know if I read your latest changes correctly, but it looks like some stuff may not be properly placed. 
1) both balloon/service module and Cryo/ballooncryo are now unlocking in the same node, while they were in different construction nodes in the original tree. Is this change intended?
2) tech levels for the RCS thrusters are now unlocking in rocketry nodes, while most of them used to unlock in avionics/flight control nodes. I think it makes sense to unlock some of the tech levels in these non-rocketry nodes.
By the way, good job rebuilding the tech tree! However, even though I did not test it, I share @Theysen concern about too many early nodes making it difficult to start over. Maybe it's worth reducing the science cost of the first tier?

**Comment by PhineasFreak** [comment]
@pap1723 Posting it here for future reference:

We can also add the default engine config in the RP-0 engine configs that have no reason to be part of RO: https://github.com/KSP-RO/RealismOverhaul/blob/master/GameData/RealismOverhaul/Engine_Configs/J2_Config.cfg#L29. We can also finally expand this to more engines that should have a different default config under RP-0.

**Comment by NathanKell** [comment]
This being now in Pap-Contracts in this repo (well, technically now merged down to master) I'm going to go ahead and close this, with much fanfare! :)

**Pull request closed without merging**